File: native-overrides/android/app/src/main/java/com/momapp/family/RevenueCatPlugin.java
package com.momapp.family;

import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import com.revenuecat.purchases.CustomerInfo;
import com.revenuecat.purchases.EntitlementInfo;
import com.revenuecat.purchases.Offering;
import com.revenuecat.purchases.Offerings;
import com.revenuecat.purchases.Package;
import com.revenuecat.purchases.Purchases;
import com.revenuecat.purchases.PurchasesConfiguration;
import com.revenuecat.purchases.PurchasesError;
import com.revenuecat.purchases.interfaces.GetOfferingsCallback;
import com.revenuecat.purchases.interfaces.LogInCallback;
import com.revenuecat.purchases.interfaces.PurchaseCallback;
import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback;
import com.revenuecat.purchases.models.StoreProduct;
import com.revenuecat.purchases.models.StoreTransaction;
import com.revenuecat.purchases.PurchaseParams;

import java.util.Map;

@CapacitorPlugin(name = "RevenueCatPlugin")
public class RevenueCatPlugin extends Plugin {
    private static final String TAG = "RevenueCatPlugin";
    private boolean isConfigured = false;

    @PluginMethod
    public void configure(PluginCall call) {
        String apiKey = call.getString("apiKey");
        if (apiKey == null) {
            call.reject("Missing apiKey");
            return;
        }

        if (isConfigured) {
            JSObject ret = new JSObject();
            ret.put("success", true);
            ret.put("alreadyConfigured", true);
            call.resolve(ret);
            return;
        }

        try {
            PurchasesConfiguration config = new PurchasesConfiguration.Builder(getContext(), apiKey).build();
            Purchases.configure(config);
            isConfigured = true;
            Log.d(TAG, "RevenueCat configured");
            JSObject ret = new JSObject();
            ret.put("success", true);
            call.resolve(ret);
        } catch (Exception e) {
            Log.e(TAG, "Configure error: " + e.getMessage());
            call.reject("Failed to configure: " + e.getMessage());
        }
    }

    @PluginMethod
    public void logIn(PluginCall call) {
        String appUserID = call.getString("appUserID");
        if (appUserID == null) {
            call.reject("Missing appUserID");
            return;
        }

        Purchases.getSharedInstance().logIn(appUserID, new LogInCallback() {
            @Override
            public void onReceived(CustomerInfo customerInfo, boolean created) {
                JSObject ret = new JSObject();
                ret.put("success", true);
                ret.put("created", created);
                ret.put("customerInfo", serializeCustomerInfo(customerInfo));
                call.resolve(ret);
            }

            @Override
            public void onError(PurchasesError error) {
                Log.e(TAG, "logIn error: " + error.getMessage());
                call.reject(error.getMessage());
            }
        });
    }

    @PluginMethod
    public void logOut(PluginCall call) {
        Purchases.getSharedInstance().logOut(new ReceiveCustomerInfoCallback() {
            @Override
            public void onReceived(CustomerInfo customerInfo) {
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }

            @Override
            public void onError(PurchasesError error) {
                call.reject(error.getMessage());
            }
        });
    }

    @PluginMethod
    public void getOfferings(PluginCall call) {
        Purchases.getSharedInstance().getOfferings(new GetOfferingsCallback() {
            @Override
            public void onReceived(Offerings offerings) {
                JSArray packages = new JSArray();
                for (Map.Entry<String, Offering> entry : offerings.getAll().entrySet()) {
                    for (Package pkg : entry.getValue().getAvailablePackages()) {
                        StoreProduct product = pkg.getProduct();
                        JSObject pkgObj = new JSObject();
                        pkgObj.put("identifier", pkg.getIdentifier());
                        pkgObj.put("packageType", pkg.getPackageType().toString());
                        pkgObj.put("productIdentifier", product.getId());
                        pkgObj.put("localizedTitle", product.getName());
                        pkgObj.put("localizedDescription", product.getDescription() != null ? product.getDescription() : "");
                        pkgObj.put("priceString", product.getPrice().getFormatted());
                        pkgObj.put("price", product.getPrice().getAmountMicros() / 1_000_000.0);
                        packages.put(pkgObj);
                    }
                }
                JSObject ret = new JSObject();
                ret.put("packages", packages);
                call.resolve(ret);
            }

            @Override
            public void onError(PurchasesError error) {
                Log.e(TAG, "getOfferings error: " + error.getMessage());
                call.reject(error.getMessage());
            }
        });
    }

    @PluginMethod
    public void purchasePackage(PluginCall call) {
        String productIdentifier = call.getString("productIdentifier");
        if (productIdentifier == null) {
            call.reject("Missing productIdentifier");
            return;
        }

        Purchases.getSharedInstance().getOfferings(new GetOfferingsCallback() {
            @Override
            public void onReceived(Offerings offerings) {
                Package foundPackage = null;
                outer:
                for (Map.Entry<String, Offering> entry : offerings.getAll().entrySet()) {
                    for (Package pkg : entry.getValue().getAvailablePackages()) {
                        if (pkg.getProduct().getId().equals(productIdentifier)) {
                            foundPackage = pkg;
                            break outer;
                        }
                    }
                }

                if (foundPackage == null) {
                    call.reject("Package not found for product: " + productIdentifier);
                    return;
                }

                final Package packageToPurchase = foundPackage;
                PurchaseParams params = new PurchaseParams.Builder(getActivity(), packageToPurchase).build();
                Purchases.getSharedInstance().purchase(params, new PurchaseCallback() {
                    @Override
                    public void onCompleted(StoreTransaction storeTransaction, CustomerInfo customerInfo) {
                        JSObject ret = new JSObject();
                        ret.put("success", true);
                        ret.put("cancelled", false);
                        ret.put("customerInfo", serializeCustomerInfo(customerInfo));
                        call.resolve(ret);
                    }

                    @Override
                    public void onError(PurchasesError error, boolean userCancelled) {
                        if (userCancelled) {
                            JSObject ret = new JSObject();
                            ret.put("success", false);
                            ret.put("cancelled", true);
                            call.resolve(ret);
                        } else {
                            Log.e(TAG, "Purchase error: " + error.getMessage());
                            call.reject(error.getMessage());
                        }
                    }
                });
            }

            @Override
            public void onError(PurchasesError error) {
                call.reject(error.getMessage());
            }
        });
    }

    @PluginMethod
    public void restorePurchases(PluginCall call) {
        Purchases.getSharedInstance().restorePurchases(new ReceiveCustomerInfoCallback() {
            @Override
            public void onReceived(CustomerInfo customerInfo) {
                JSObject ret = new JSObject();
                ret.put("success", true);
                ret.put("customerInfo", serializeCustomerInfo(customerInfo));
                call.resolve(ret);
            }

            @Override
            public void onError(PurchasesError error) {
                Log.e(TAG, "Restore error: " + error.getMessage());
                call.reject(error.getMessage());
            }
        });
    }

    @PluginMethod
    public void getCustomerInfo(PluginCall call) {
        Purchases.getSharedInstance().getCustomerInfo(new ReceiveCustomerInfoCallback() {
            @Override
            public void onReceived(CustomerInfo customerInfo) {
                JSObject ret = new JSObject();
                ret.put("success", true);
                ret.put("customerInfo", serializeCustomerInfo(customerInfo));
                call.resolve(ret);
            }

            @Override
            public void onError(PurchasesError error) {
                call.reject(error.getMessage());
            }
        });
    }

    private JSObject serializeCustomerInfo(CustomerInfo info) {
        JSObject result = new JSObject();
        if (info == null) return result;

        JSArray activeEntitlements = new JSArray();
        for (Map.Entry<String, EntitlementInfo> entry : info.getEntitlements().getActive().entrySet()) {
            activeEntitlements.put(entry.getKey());
        }
        result.put("activeEntitlements", activeEntitlements);

        JSArray activeSubscriptions = new JSArray();
        for (String sub : info.getActiveSubscriptions()) {
            activeSubscriptions.put(sub);
        }
        result.put("activeSubscriptions", activeSubscriptions);

        JSArray entitlementDetails = new JSArray();
        for (Map.Entry<String, EntitlementInfo> entry : info.getEntitlements().getAll().entrySet()) {
            EntitlementInfo entitlement = entry.getValue();
            JSObject detail = new JSObject();
            detail.put("identifier", entry.getKey());
            detail.put("isActive", entitlement.isActive());
            detail.put("productIdentifier", entitlement.getProductIdentifier());
            detail.put("willRenew", entitlement.getWillRenew());
            if (entitlement.getExpirationDate() != null) {
                detail.put("expirationDate", entitlement.getExpirationDate().toString());
            }
            entitlementDetails.put(detail);
        }
        result.put("entitlements", entitlementDetails);

        return result;
    }
}

File: native-overrides/android/app/src/main/java/com/momapp/family/RevenueCatPlugin.kt
package com.momapp.family

import android.util.Log
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.EntitlementInfo
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Offerings
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesError
import com.revenuecat.purchases.interfaces.GetOfferingsCallback
import com.revenuecat.purchases.interfaces.LogInCallback
import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
import com.revenuecat.purchases.models.StoreTransaction

@CapacitorPlugin(name = "RevenueCatPlugin")
class RevenueCatPlugin : Plugin() {

    companion object {
        private const val TAG = "RevenueCatPlugin"
    }

    private var isConfigured = false

    @PluginMethod
    fun configure(call: PluginCall) {
        val apiKey = call.getString("apiKey") ?: run {
            call.reject("Missing apiKey")
            return
        }

        if (isConfigured) {
            call.resolve(JSObject().also {
                it.put("success", true)
                it.put("alreadyConfigured", true)
            })
            return
        }

        try {
            Purchases.configure(PurchasesConfiguration.Builder(context, apiKey).build())
            isConfigured = true
            Log.d(TAG, "RevenueCat configured successfully")
            call.resolve(JSObject().also { it.put("success", true) })
        } catch (e: Exception) {
            Log.e(TAG, "Configure error: ${e.message}")
            call.reject("Failed to configure: ${e.message}")
        }
    }

    @PluginMethod
    fun logIn(call: PluginCall) {
        val appUserID = call.getString("appUserID") ?: run {
            call.reject("Missing appUserID")
            return
        }

        Purchases.sharedInstance.logIn(appUserID, object : LogInCallback {
            override fun onReceived(customerInfo: CustomerInfo, created: Boolean) {
                call.resolve(JSObject().also { ret ->
                    ret.put("success", true)
                    ret.put("created", created)
                    ret.put("customerInfo", serializeCustomerInfo(customerInfo))
                })
            }

            override fun onError(error: PurchasesError) {
                Log.e(TAG, "logIn error: ${error.message}")
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun logOut(call: PluginCall) {
        Purchases.sharedInstance.logOut(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                call.resolve(JSObject().also { it.put("success", true) })
            }

            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun getOfferings(call: PluginCall) {
        Purchases.sharedInstance.getOfferings(object : GetOfferingsCallback {
            override fun onReceived(offerings: Offerings) {
                val packages = JSArray()
                for ((_, offering) in offerings.all) {
                    for (pkg in offering.availablePackages) {
                        val pkgObj = JSObject().also { obj ->
                            obj.put("identifier", pkg.identifier)
                            obj.put("packageType", pkg.packageType.toString())
                            obj.put("productIdentifier", pkg.product.id)
                            obj.put("localizedTitle", pkg.product.name)
                            obj.put("localizedDescription", pkg.product.description ?: "")
                            obj.put("priceString", pkg.product.price.formatted)
                            obj.put("price", pkg.product.price.amountMicros / 1_000_000.0)
                        }
                        packages.put(pkgObj)
                    }
                }
                call.resolve(JSObject().also { it.put("packages", packages) })
            }

            override fun onError(error: PurchasesError) {
                Log.e(TAG, "getOfferings error: ${error.message}")
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun purchasePackage(call: PluginCall) {
        val productIdentifier = call.getString("productIdentifier") ?: run {
            call.reject("Missing productIdentifier")
            return
        }

        Purchases.sharedInstance.getOfferings(object : GetOfferingsCallback {
            override fun onReceived(offerings: Offerings) {
                var foundPackage: Package? = null
                loop@ for ((_, offering) in offerings.all) {
                    for (pkg in offering.availablePackages) {
                        if (pkg.product.id == productIdentifier) {
                            foundPackage = pkg
                            break@loop
                        }
                    }
                }

                val packageToPurchase = foundPackage ?: run {
                    call.reject("Package not found for product: $productIdentifier")
                    return
                }

                Purchases.sharedInstance.purchase(
                    PurchaseParams.Builder(activity, packageToPurchase).build()
                ) { _: StoreTransaction?, customerInfo: CustomerInfo?, error: PurchasesError?, userCancelled: Boolean ->
                    when {
                        userCancelled -> call.resolve(JSObject().also {
                            it.put("success", false)
                            it.put("cancelled", true)
                        })
                        error != null -> {
                            Log.e(TAG, "Purchase error: ${error.message}")
                            call.reject(error.message)
                        }
                        else -> call.resolve(JSObject().also {
                            it.put("success", true)
                            it.put("cancelled", false)
                            it.put("customerInfo", serializeCustomerInfo(customerInfo))
                        })
                    }
                }
            }

            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun restorePurchases(call: PluginCall) {
        Purchases.sharedInstance.restorePurchases(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                call.resolve(JSObject().also {
                    it.put("success", true)
                    it.put("customerInfo", serializeCustomerInfo(customerInfo))
                })
            }

            override fun onError(error: PurchasesError) {
                Log.e(TAG, "Restore error: ${error.message}")
                call.reject(error.message)
            }
        })
    }

    @PluginMethod
    fun getCustomerInfo(call: PluginCall) {
        Purchases.sharedInstance.getCustomerInfo(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                call.resolve(JSObject().also {
                    it.put("success", true)
                    it.put("customerInfo", serializeCustomerInfo(customerInfo))
                })
            }

            override fun onError(error: PurchasesError) {
                call.reject(error.message)
            }
        })
    }

    private fun serializeCustomerInfo(info: CustomerInfo?): JSObject {
        val result = JSObject()
        info ?: return result

        val activeEntitlements = JSArray()
        for (key in info.entitlements.active.keys) {
            activeEntitlements.put(key)
        }
        result.put("activeEntitlements", activeEntitlements)

        val activeSubscriptions = JSArray()
        for (sub in info.activeSubscriptions) {
            activeSubscriptions.put(sub)
        }
        result.put("activeSubscriptions", activeSubscriptions)

        val entitlementDetails = JSArray()
        for ((key, entitlement) in info.entitlements.all) {
            val detail = JSObject().also { d ->
                d.put("identifier", key)
                d.put("isActive", entitlement.isActive)
                d.put("productIdentifier", entitlement.productIdentifier)
                d.put("willRenew", entitlement.willRenew)
                entitlement.expirationDate?.let { d.put("expirationDate", it.toString()) }
            }
            entitlementDetails.put(detail)
        }
        result.put("entitlements", entitlementDetails)

        return result
    }
}
